File: targets/java-web/src/main/java/com/example/vulnapp/vulns/Vuln02SqlLogin.java
package com.example.vulnapp.vulns;

import com.example.vulnapp.infra.Db;
import com.example.vulnapp.vulns.crypto.WeakHash;
import org.springframework.web.bind.annotation.*;
import java.sql.*;
import java.util.*;

@RestController
public class Vuln02SqlLogin {
    private final Db db;
    public Vuln02SqlLogin(Db db) { this.db = db; }

    @PostMapping("/vuln02/login")
    public Map<String,Object> login(@RequestParam String username, @RequestParam String password) throws Exception {
        Map<String,Object> out = new HashMap<>();
        try (Connection c = db.getConnection(); Statement st = c.createStatement()) {
            String hash = WeakHash.md5(password);
            // VULN:VULN-02:CWE-89:taint SQL injection in authentication (login bypass)
            String sql = "SELECT id, username, role FROM users WHERE username = '" + username + "' AND password = '" + hash + "'";
            ResultSet rs = st.executeQuery(sql);
            if (rs.next()) { out.put("status","ok"); out.put("user", rs.getString("username")); out.put("role", rs.getString("role")); }
            else out.put("status","invalid credentials");
        }
        return out;
    }
}

File: targets/kotlin-web/src/main/kotlin/com/example/vulnapp/vulns/Vuln02SqlLogin.kt
package com.example.vulnapp.vulns

import com.example.vulnapp.infra.Db
import org.springframework.web.bind.annotation.*


@RestController
class Vuln02SqlLogin(private val db: Db) {

    @PostMapping("/vuln02/login")
    fun login(@RequestParam username: String, @RequestParam password: String): Map<String, Any?> {
        db.connection().use { c -> c.createStatement().use { st ->
            val hash = com.example.vulnapp.vulns.crypto.WeakHash.md5(password)
            // VULN:VULN-02:CWE-89:taint SQL injection in authentication (login bypass)
            val rs = st.executeQuery("SELECT id, username, role FROM users WHERE username = '$username' AND password = '$hash'")
            return if (rs.next()) mapOf("status" to "ok", "user" to rs.getString("username"), "role" to rs.getString("role"))
                   else mapOf("status" to "invalid")
        } }
    }
}
